docs/en/StreamFusion/config_advanced.md
# Advanced Configuration of StreamFusion

This section details all the environment variables that can be configured to customize StreamFusion's behavior. These variables can be passed either as Docker secrets or as container environment variables (in uppercase).

!!! danger "Caution regarding automatic configuration"
    Some environment variables are automatically configured when the application starts, based on your installation. Manually overriding these variables can lead to undesired behavior of StreamFusion, particularly regarding interactions with debridging services.

    The concerned variables include:

    - `PROXIED_LINK`
    - `RD_UNIQUE_ACCOUNT`
    - `AD_UNIQUE_ACCOUNT`
    - `AD_USE_PROXY`
    - `JACKETT_ENABLE`
    - `YGG_UNIQUE_ACCOUNT`
    - `SHAREWOOD_UNIQUE_ACCOUNT`

    Before manually modifying these variables, make sure you understand their impact on the application's functioning. Incorrect configuration could compromise the stability and security of your StreamFusion installation.

## General Variables

| Variable | Description | Default Value |
|----------|-------------|----------------|
| `SESSION_KEY` | Session key for the application | Automatically generated |
| `USE_HTTPS` | Enable HTTPS | `false` |
| `DEFAULT_DEBRID_SERVICE` | Default debridging service | `RD` (RealDebrid) |

!!! info "Session Key"
    The `SESSION_KEY` is automatically generated when the container starts, ensuring a unique key for each StreamFusion installation. This practice enhances the security of your application.

    - It is strongly advised against manually setting this value via an environment variable.
    - The automatic rotation of this key will not affect existing user connections.
    - This approach follows best practices for web application security.

## Server Configuration

| Variable | Description | Default Value |
|----------|-------------|----------------|
| `WORKERS_COUNT` | Number of workers for the application | Automatically calculated |
| `PORT` | Port on which the application listens | `8080` |
| `HOST` | Host on which the application listens | `0.0.0.0` |
| `GUNICORN_TIMEOUT` | Timeout for Gunicorn (in seconds) | `180` |
| `AIOHTTP_TIMEOUT` | Timeout for aiohttp (in seconds) | `7200` |

## Proxy Configuration

| Variable | Description | Default Value |
|----------|-------------|----------------|
| `PROXIED_LINK` | Enable link proxification | Depends on configuration |
| `PROXY_URL` | URL of the proxy to use | `None` |
| `PLAYBACK_PROXY` | Use proxy for stream links from the server | `None` |

!!! info "Functioning of proxy variables"
    - `PROXIED_LINK`: Transforms the application into a proxy, passing all streams through the server. Used to share a debridger account among all users.
    - `PROXY_URL`: Contains the URL of the user's proxy.
    - `PLAYBACK_PROXY`: Applies the proxy to all StreamFusion interactions with debridgers, including streaming links. Useful when `PROXIED_LINK` is enabled, so that video streaming also passes through the configured proxy.

!!! warning "Use of proxies and interactions with debridgers"
    It's important to understand the different aspects of interactions with debridgers:

    1. **Streaming**: Concerns the streaming links of debrided content.
    2. **API Interactions**: Concerns requests to the debridgers' API.

    ---

    - Some debridgers (like AllDebrid) block API requests coming from servers, hence the use of a proxy to contact them.
    - It may happen that server IP addresses are banned by RealDebrid or AllDebrid, preventing even the streaming of video links. In this case, activate `PLAYBACK_PROXY` to pass video playback through the proxy.
    - The use of `PLAYBACK_PROXY` is generally avoided to save resources and reduce latency, except when necessary.

## RealDebrid Configuration

| Variable | Description | Default Value |
|----------|-------------|----------------|
| `RD_TOKEN` | RealDebrid authentication token | `None` |
| `RD_UNIQUE_ACCOUNT` | Use a unique RealDebrid account | Depends on configuration |

## AllDebrid Configuration

| Variable | Description | Default Value |
|----------|-------------|----------------|
| `AD_TOKEN` | AllDebrid authentication token | `None` |
| `AD_UNIQUE_ACCOUNT` | Use a unique AllDebrid account | Depends on configuration |
| `AD_USER_APP` | Application name for AllDebrid | `streamfusion` |
| `AD_USER_IP` | User IP for AllDebrid | `None` |
| `AD_USE_PROXY` | Use a proxy for AllDebrid | Depends on configuration |

## Logging Configuration

| Variable | Description | Default Value |
|----------|-------------|----------------|
| `LOG_LEVEL` | Log level | `INFO` |
| `LOG_PATH` | Path to the log file | `/app/config/logs/stream-fusion.log` |
| `LOG_REDACTED` | Hide sensitive information in logs | `true` |

## Security Configuration

| Variable | Description | Default Value |
|----------|-------------|----------------|
| `SECRET_API_KEY` | Secret API key | `None` |
| `SECURITY_HIDE_DOCS` | Hide API documentation | `true` |

## PostgreSQL Configuration

| Variable | Description | Default Value |
|----------|-------------|----------------|
| `PG_HOST` | PostgreSQL host | `stremio-postgres` |
| `PG_PORT` | PostgreSQL port | `5432` |
| `PG_USER` | PostgreSQL user | `streamfusion` |
| `PG_PASS` | PostgreSQL password | `streamfusion` |
| `PG_BASE` | PostgreSQL database name | `streamfusion` |
| `PG_ECHO` | Enable SQL echo | `false` |

## Redis Configuration

| Variable | Description | Default Value |
|----------|-------------|----------------|
| `REDIS_HOST` | Redis host | `redis` |
| `REDIS_PORT` | Redis port | `6379` |
| `REDIS_DB` | Redis database number | `5` |
| `REDIS_EXPIRATION` | Redis key expiration duration (in seconds) | `604800` |
| `REDIS_PASSWORD` | Redis password | `None` |

## TMDB Configuration

| Variable | Description | Default Value |
|----------|-------------|----------------|
| `TMDB_API_KEY` | TMDB API key | `None` |

## Jackett Configuration

| Variable | Description | Default Value |
|----------|-------------|----------------|
| `JACKETT_HOST` | Jackett host | `jackett` |
| `JACKETT_SCHEMA` | Jackett URL schema | `http` |
| `JACKETT_PORT` | Jackett port | `9117` |
| `JACKETT_API_KEY` | Jackett API key | `None` |
| `JACKETT_ENABLE` | Enable Jackett integration | Depends on configuration |

## Zilean DMM API Configuration

| Variable | Description | Default Value |
|----------|-------------|----------------|
| `ZILEAN_HOST` | Zilean host | `zilean` |
| `ZILEAN_PORT` | Zilean port | `8181` |
| `ZILEAN_SCHEMA` | Zilean URL schema | `http` |
| `ZILEAN_MAX_WORKERS` | Maximum number of Zilean workers | `4` |
| `ZILEAN_POOL_CONNECTIONS` | Zilean connection pool size | `10` |
| `ZILEAN_API_POOL_MAXSIZE` | Zilean API pool maximum size | `10` |
| `ZILEAN_MAX_RETRY` | Maximum number of Zilean retries | `3` |

## YGGFlix Configuration

| Variable | Description | Default Value |
|----------|-------------|----------------|
| `YGGFLIX_URL` | YGGFlix URL | `https://yggflix.fr` |
| `YGGFLIX_MAX_WORKERS` | Maximum number of YGGFlix workers | `4` |
| `YGG_PASSKEY` | YGG passkey | `None` |
| `YGG_UNIQUE_ACCOUNT` | Use a unique YGG account | Depends on configuration |

## Sharewood Configuration

| Variable | Description | Default Value |
|----------|-------------|----------------|
| `SHAREWOOD_URL` | Sharewood URL | `https://www.sharewood.tv` |
| `SHAREWOOD_MAX_WORKERS` | Maximum number of Sharewood workers | `4` |
| `SHAREWOOD_PASSKEY` | Sharewood passkey | `None` |
| `SHAREWOOD_UNIQUE_ACCOUNT` | Use a unique Sharewood account | Depends on configuration |

## Public Cache Configuration

| Variable | Description | Default Value |
|----------|-------------|----------------|
| `PUBLIC_CACHE_URL` | Public cache URL | `https://stremio-jackett-cacher.elfhosted.com/` |

## Development Configuration

| Variable | Description | Default Value |
|----------|-------------|----------------|
| `DEBUG` | Enable debug mode | `false` |
| `DEV_HOST` | Development host | `0.0.0.0` |
| `DEV_PORT` | Development port | `8080` |
| `DEVELOP` | Enable development mode | `false` |
| `RELOAD` | Enable automatic reloading | `false` |

!!! warning "Security"
    Make sure not to expose sensitive information such as tokens and API keys in unsecured environments.

docs/fr/StreamFusion/config_advanced.md
# Configuration avancée de StreamFusion

Cette section détaille toutes les variables d'environnement qui peuvent être configurées pour personnaliser le comportement de StreamFusion. Ces variables peuvent être passées soit en tant que secrets Docker, soit en tant que variables d'environnement du conteneur (en majuscules).

!!! danger "Attention à la configuration automatique"
    Certaines variables d'environnement se configurent automatiquement au démarrage de l'application en fonction de votre installation. Surcharger ces variables manuellement peut entraîner des comportements indésirables de StreamFusion, particulièrement en ce qui concerne les interactions avec les services de débridage.

    Les variables concernées incluent notamment :

    - `PROXIED_LINK`
    - `RD_UNIQUE_ACCOUNT`
    - `AD_UNIQUE_ACCOUNT`
    - `AD_USE_PROXY`
    - `JACKETT_ENABLE`
    - `YGG_UNIQUE_ACCOUNT`
    - `SHAREWOOD_UNIQUE_ACCOUNT`

    Avant de modifier manuellement ces variables, assurez-vous de bien comprendre leur impact sur le fonctionnement de l'application. Une configuration incorrecte pourrait compromettre la stabilité et la sécurité de votre installation StreamFusion.

## Variables générales

| Variable | Description | Valeur par défaut |
|----------|-------------|-------------------|
| `SESSION_KEY` | Clé de session pour l'application | Générée automatiquement |
| `USE_HTTPS` | Activer HTTPS | `false` |
| `DEFAULT_DEBRID_SERVICE` | Service de débridage par défaut | `RD` (RealDebrid) |

!!! info "Session Key"
    La `SESSION_KEY` est générée automatiquement au démarrage du conteneur, garantissant ainsi une clé unique pour chaque installation de StreamFusion. Cette pratique renforce la sécurité de votre application.

    - Il est fortement déconseillé de fixer manuellement cette valeur via une variable d'environnement.
    - La rotation automatique de cette clé n'affectera pas les connexions utilisateurs existantes.
    - Cette approche suit les meilleures pratiques en matière de sécurité pour les applications web.

## Configuration du serveur

| Variable | Description | Valeur par défaut |
|----------|-------------|-------------------|
| `WORKERS_COUNT` | Nombre de workers pour l'application | Calculé automatiquement |
| `PORT` | Port sur lequel l'application écoute | `8080` |
| `HOST` | Hôte sur lequel l'application écoute | `0.0.0.0` |
| `GUNICORN_TIMEOUT` | Timeout pour Gunicorn (en secondes) | `180` |
| `AIOHTTP_TIMEOUT` | Timeout pour aiohttp (en secondes) | `7200` |

## Configuration du proxy

| Variable | Description | Valeur par défaut |
|----------|-------------|-------------------|
| `PROXIED_LINK` | Activer la proxification des liens | Dépend de la configuration |
| `PROXY_URL` | URL du proxy à utiliser | `None` |
| `PLAYBACK_PROXY` | Utiliser le proxy pour les lien de stream depuis le serveur | `None` |


!!! info "Fonctionnement des variables de proxy"
    - `PROXIED_LINK` : Transforme l'application en proxy, faisant passer tous les streams par le serveur. Utilisé pour partager un compte de débrideur entre tous les utilisateurs.
    - `PROXY_URL` : Contient l'URL du proxy de l'utilisateur.
    - `PLAYBACK_PROXY` : Applique le proxy à toutes les interactions de StreamFusion avec les débrideurs, y compris les liens de streaming. Utile quand `PROXIED_LINK` est activé, pour que le streaming vidéo passe également par le proxy configuré.

!!! warning "Utilisation des proxys et interactions avec les débrideurs"
    Il est important de comprendre les différents aspects des interactions avec les débrideurs:

    1. **Streaming** : Concerne les liens de streaming des contenus débridés.
    2. **Interactions API** : Concerne les requêtes à l'API des débrideurs.

    ---

    - Certains débrideurs (comme AllDebrid) bloquent les requêtes API provenant de serveurs, d'où l'utilisation d'un proxy pour les contacter.
    - Il peut arriver que des adresses IP de serveurs soient bannies par RealDebrid ou AllDebrid, empêchant même le streaming de liens vidéo. Dans ce cas, activez `PLAYBACK_PROXY` pour faire passer les lectures vidéo par le proxy.
    - L'utilisation de `PLAYBACK_PROXY` est généralement évitée pour économiser les ressources et réduire la latence, sauf en cas de nécessité.

## Configuration de RealDebrid

| Variable | Description | Valeur par défaut |
|----------|-------------|-------------------|
| `RD_TOKEN` | Token d'authentification RealDebrid | `None` |
| `RD_UNIQUE_ACCOUNT` | Utiliser un compte RealDebrid unique | Dépend de la configuration |

## Configuration d'AllDebrid

| Variable | Description | Valeur par défaut |
|----------|-------------|-------------------|
| `AD_TOKEN` | Token d'authentification AllDebrid | `None` |
| `AD_UNIQUE_ACCOUNT` | Utiliser un compte AllDebrid unique | Dépend de la configuration |
| `AD_USER_APP` | Nom de l'application pour AllDebrid | `streamfusion` |
| `AD_USER_IP` | IP de l'utilisateur pour AllDebrid | `None` |
| `AD_USE_PROXY` | Utiliser un proxy pour AllDebrid | Dépend de la configuration |

## Configuration Logging

| Variable | Description | Valeur par défaut |
|----------|-------------|-------------------|
| `LOG_LEVEL` | Niveau de log | `INFO` |
| `LOG_PATH` | Chemin du fichier de log | `/app/config/logs/stream-fusion.log` |
| `LOG_REDACTED` | Masquer les informations sensibles dans les logs | `true` |

## Configuration Sécurité

| Variable | Description | Valeur par défaut |
|----------|-------------|-------------------|
| `SECRET_API_KEY` | Clé API secrète | `None` |
| `SECURITY_HIDE_DOCS` | Cacher la documentation de l'API | `true` |

## Configuration PostgreSQL

| Variable | Description | Valeur par défaut |
|----------|-------------|-------------------|
| `PG_HOST` | Hôte PostgreSQL | `stremio-postgres` |
| `PG_PORT` | Port PostgreSQL | `5432` |
| `PG_USER` | Utilisateur PostgreSQL | `streamfusion` |
| `PG_PASS` | Mot de passe PostgreSQL | `streamfusion` |
| `PG_BASE` | Nom de la base de données PostgreSQL | `streamfusion` |
| `PG_ECHO` | Activer l'echo SQL | `false` |

## Configuration Redis

| Variable | Description | Valeur par défaut |
|----------|-------------|-------------------|
| `REDIS_HOST` | Hôte Redis | `redis` |
| `REDIS_PORT` | Port Redis | `6379` |
| `REDIS_DB` | Numéro de la base de données Redis | `5` |
| `REDIS_EXPIRATION` | Durée d'expiration des clés Redis (en secondes) | `604800` |
| `REDIS_PASSWORD` | Mot de passe Redis | `None` |

## Configuration TMDB

| Variable | Description | Valeur par défaut |
|----------|-------------|-------------------|
| `TMDB_API_KEY` | Clé API TMDB | `None` |

## Configuration Jackett

| Variable | Description | Valeur par défaut |
|----------|-------------|-------------------|
| `JACKETT_HOST` | Hôte Jackett | `jackett` |
| `JACKETT_SCHEMA` | Schéma de l'URL Jackett | `http` |
| `JACKETT_PORT` | Port Jackett | `9117` |
| `JACKETT_API_KEY` | Clé API Jackett | `None` |
| `JACKETT_ENABLE` | Activer l'intégration Jackett | Dépend de la configuration |

## Configuration Zilean DMM API

| Variable | Description | Valeur par défaut |
|----------|-------------|-------------------|
| `ZILEAN_HOST` | Hôte Zilean | `zilean` |
| `ZILEAN_PORT` | Port Zilean | `8181` |
| `ZILEAN_SCHEMA` | Schéma de l'URL Zilean | `http` |
| `ZILEAN_MAX_WORKERS` | Nombre maximum de workers Zilean | `4` |
| `ZILEAN_POOL_CONNECTIONS` | Taille du pool de connexions Zilean | `10` |
| `ZILEAN_API_POOL_MAXSIZE` | Taille maximale du pool API Zilean | `10` |
| `ZILEAN_MAX_RETRY` | Nombre maximum de tentatives Zilean | `3` |

## Configuration YGGFlix

| Variable | Description | Valeur par défaut |
|----------|-------------|-------------------|
| `YGGFLIX_URL` | URL de YGGFlix | `https://yggflix.fr` |
| `YGGFLIX_MAX_WORKERS` | Nombre maximum de workers YGGFlix | `4` |
| `YGG_PASSKEY` | Passkey YGG | `None` |
| `YGG_UNIQUE_ACCOUNT` | Utiliser un compte YGG unique | Dépend de la configuration |

## Configuration Sharewood

| Variable | Description | Valeur par défaut |
|----------|-------------|-------------------|
| `SHAREWOOD_URL` | URL de Sharewood | `https://www.sharewood.tv` |
| `SHAREWOOD_MAX_WORKERS` | Nombre maximum de workers Sharewood | `4` |
| `SHAREWOOD_PASSKEY` | Passkey Sharewood | `None` |
| `SHAREWOOD_UNIQUE_ACCOUNT` | Utiliser un compte Sharewood unique | Dépend de la configuration |

## Configuration Cache public

| Variable | Description | Valeur par défaut |
|----------|-------------|-------------------|
| `PUBLIC_CACHE_URL` | URL du cache public | `https://stremio-jackett-cacher.elfhosted.com/` |

## Configuration de développement

| Variable | Description | Valeur par défaut |
|----------|-------------|-------------------|
| `DEBUG` | Activer le mode debug | `false` |
| `DEV_HOST` | Hôte de développement | `0.0.0.0` |
| `DEV_PORT` | Port de développement | `8080` |
| `DEVELOP` | Activer le mode développement | `false` |
| `RELOAD` | Activer le rechargement automatique | `false` |

!!! warning "Sécurité"
    Assurez-vous de ne pas exposer les informations sensibles telles que les tokens et les clés API dans des environnements non sécurisés.




stream_fusion/settings.py
import enum
import multiprocessing
import os
from pydantic import Field, ValidationError, field_validator
from pydantic_settings import BaseSettings, SettingsConfigDict
from yarl import URL


class LogLevel(str, enum.Enum):
    """Possible log levels."""

    NOTSET = "NOTSET"
    DEBUG = "DEBUG"
    INFO = "INFO"
    WARNING = "WARNING"
    ERROR = "ERROR"
    FATAL = "FATAL"

class DebridService(str, enum.Enum):
    """Possible debrid services."""

    RD = "RD"
    AD = "AD"

def get_default_worker_count():
    """
    Calculate the default number of workers based on CPU cores.
    Returns the number of CPU cores multiplied by 2, with a minimum of 2 and a maximum of 6.
    """
    return min(max(multiprocessing.cpu_count() * 2, 2), 6)

def check_env_variable(var_name):
    value = os.getenv(var_name.upper())
    
    if value and isinstance(value, str) and len(value.strip()) >= 10:
        return True
    return False

class Settings(BaseSettings):
    """Settings for the application"""

    # STREAM-FUSION
    version_path: str = "/app/pyproject.toml"
    workers_count: int = get_default_worker_count()
    port: int = 8080
    host: str = "0.0.0.0"
    gunicorn_timeout: int = 180
    aiohttp_timeout: int = 7200
    session_key: str = Field(
        default_factory=lambda: os.getenv(
            "SESSION_KEY",
            "331cbfe48117fcba53d09572b10d2fc293d86131dc51be46d8aa9843c2e9f48d",
        )
    )
    use_https: bool = False
    default_debrid_service: DebridService = DebridService.RD

    # PROXY
    proxied_link: bool = check_env_variable("RD_TOKEN") or check_env_variable("AD_TOKEN")
    proxy_url: str | URL | None = None
    playback_proxy: bool | None = (
        None  # If set, the link will be proxied through the given proxy.
    )

    # REALDEBRID
    rd_token: str | None = None
    rd_unique_account: bool = check_env_variable("RD_TOKEN")
    rd_base_url: str = "https://api.real-debrid.com/rest"
    rd_api_version: str = "1.0"

    # ALLDEBRID
    ad_token: str | None = None
    ad_unique_account: bool = check_env_variable("AD_TOKEN")
    ad_user_app: str = "streamfusion"
    ad_user_ip: str | None = None
    ad_use_proxy: bool = check_env_variable("PROXY_URL")
    ad_base_url: str = "https://api.alldebrid.com"
    ad_api_version: str = "v4"

    # TORBOX
    tb_token: str | None = None
    tb_unique_account: bool = check_env_variable("TB_TOKEN")
    tb_base_url: str = "https://api.torbox.app"
    tb_api_version: str = "v1"

    # LOGGING
    log_level: LogLevel = LogLevel.INFO
    log_path: str = "/app/config/logs/stream-fusion.log"
    log_redacted: bool = True

    # SECUIRITY
    secret_api_key: str | None = None
    security_hide_docs: bool = True

    # POSTGRESQL_DB 
    # TODO: Change the values, but break dev environment
    pg_host: str = "stremio-postgres"
    pg_port: int = 5432
    pg_user: str = "streamfusion" #"stremio"
    pg_pass: str = "streamfusion" #"stremio"
    pg_base: str = "streamfusion"
    pg_echo: bool = False

    # REDIS
    redis_host: str = "redis"
    redis_port: int = 6379
    redis_db: int = 5
    redis_expiration: int = 604800
    redis_password: str | None = None

    # TMDB
    tmdb_api_key: str | None = None

    # JACKETT
    jackett_host: str = "jackett"
    jackett_schema: str = "http"
    jackett_port: int = 9117
    jackett_api_key: str | None = None
    jackett_enable: bool = check_env_variable("JACKETT_API_KEY")

    # ZILEAN DMM API
    zilean_host: str = "zilean"
    zilean_port: int = 8181
    zilean_schema: str = "http"
    zilean_max_workers: int = 4
    zilean_pool_connections: int = 10
    zilean_api_pool_maxsize: int = 10
    zilean_max_retry: int = 3

    # YGGFLIX
    yggflix_url: str = "https://yggflix.fr"
    yggflix_max_workers: int = 4
    ygg_passkey: str | None = None
    ygg_unique_account: bool = check_env_variable("YGG_PASSKEY")

    # SHAREWOOD
    sharewood_url: str = "https://www.sharewood.tv"
    sharewood_max_workers: int = 4
    sharewood_passkey: str | None = None
    sharewood_unique_account: bool = check_env_variable("SHAREWOOD_PASSKEY")

    # PUBLIC_CACHE
    public_cache_url: str = "https://stremio-jackett-cacher.elfhosted.com/"

    # DEVELOPMENT
    debug: bool = False
    dev_host: str = "0.0.0.0"
    dev_port: int = 8080
    develop: bool = False
    reload: bool = False

    @field_validator('proxy_url')
    @classmethod
    def validate_and_create_proxy_url(cls, v: str | None) -> URL | None:
        if v is None:
            return None
        
        v = v.strip('"\'')
        if not v.startswith(('http://', 'https://')):
            v = 'http://' + v
        try:
            return URL(v)
        except ValueError as e:
            raise ValueError(f"Invalid proxy URL: {e}")

    @property
    def pg_url(self) -> URL:
        """
        Assemble database URL from settings.

        :return: database URL.
        """
        return URL.build(
            scheme="postgresql+asyncpg",
            host=self.pg_host,
            port=self.pg_port,
            user=self.pg_user,
            password=self.pg_pass,
            path=f"/{self.pg_base}",
        )
    @property
    def jackett_url(self) -> URL:
        """
        Assemble Jackett URL from settings.
        :return: Jackett URL.
        """
        url = URL.build(
            scheme=self.jackett_schema,
            host=self.jackett_host,
            port=self.jackett_port,
        )
        if self.jackett_api_key:
            url = url.with_query({"apikey": self.jackett_api_key})
        return url

    @property
    def zilean_url(self) -> URL:
        """
        Assemble Zilean URL from settings.
        :return: Zilean URL.
        """
        return URL.build(
            scheme=self.zilean_schema,
            host=self.zilean_host,
            port=self.zilean_port,
        )

    @property
    def redis_url(self) -> URL:
        """
        Assemble Redis URL from settings.
        :return: Redis URL.
        """
        url = URL.build(
            scheme="redis",
            host=self.redis_host,
            port=self.redis_port,
        )
        if self.redis_password:
            url = url.with_password(self.redis_password)
        return url

    model_config = SettingsConfigDict(
        env_file=".env", secrets_dir="/run/secrets", env_file_encoding="utf-8"
    )


try:
    settings = Settings()
except ValidationError as e:
    raise RuntimeError(f"Configuration validation error: {e}")
